### PR TITLE
Prevent dead-store elimination when clearing secrets in examples

### PR DESCRIPTION
**Pull request by Harshil-Jani**
Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

### CODE REVIEW

**Comment by real-or-random** [comment]
It's probably a good idea to add this to the examples, but I think an approach that has a separate function like in https://github.com/bitcoin-core/secp256k1/pull/636/commits/494965880b47c3f2b752391b0c757614d6cb9f75 is cleaner. `secure_erase` sounds like a good function name, for example.

**Comment by alokeutpal** [comment]
Please help me a little. I want to get it in phantom wallet. One more thing
I want to know is my address in my phantom wallet where can I get it?

**Comment by alokeutpal** [comment]
Really I am non technical person only understand mobile off on help me

**Comment by Harshil-Jani** [comment]
> It's probably a good idea to add this to the examples, but I think an approach that has a separate function like in https://github.com/bitcoin-core/secp256k1/commit/494965880b47c3f2b752391b0c757614d6cb9f75 is cleaner. secure_erase sounds like a good function name, for example.

Thanks @real-or-random  for this !! Will look at this soon.

> Please help me a little. I want to get it in phantom wallet. One more thing
I want to know is my address in my phantom wallet where can I get it?

Sorry ?? I guess, This would be a wrong place for the question.

**Comment by Harshil-Jani** [comment]
@real-or-random Let me know your feedback on this. Also, I was wondering is there any header where we can club this ? or how about a seperate header for `secure_erase` instead of writing it for each file.

**Comment by real-or-random** [comment]
Yes, looks better now. Now if we include such a function, I still feel we should recommend state-of-the art practices, which is what I implemented in https://github.com/bitcoin-core/secp256k1/commit/494965880b47c3f2b752391b0c757614d6cb9f75. Note that essentially the same code was added to Bitcoin Core, see https://github.com/bitcoin-core/secp256k1/commit/494965880b47c3f2b752391b0c757614d6cb9f75.

> Also, I was wondering is there any header where we can club this ? or how about a seperate header for `secure_erase` instead of writing it for each file.

I think we could put it in `random.h` and rename this to `util.h` or `examples_util.h` to better distinguish it from our own `util.h`.

**Comment by Harshil-Jani** [comment]
@real-or-random I tried renaming the file. But, I think that would need to change few other things from `secp256k1.h` which could be part of another patch.

**Comment by real-or-random** [comment]
> @real-or-random I tried renaming the file. But, I think that would need to change few other things from `secp256k1.h` which could be part of another patch.

Why do you think so? This should be unrelated?

**Comment by Harshil-Jani** [comment]
> > @real-or-random I tried renaming the file. But, I think that would need to change few other things from `secp256k1.h` which could be part of another patch.
> 
> Why do you think so? This should be unrelated?

My bad. I just renamed it and `SECP256K1_INLINE`  instantly throwed an error of `explicit type is missing ('int' assumed)` (And I thought how would renaming a header affect the same file altough I knew we need to change it's name in other files as well)so I assumed we need to fix this from secp.h But on further inspection I got to know that after renaming the header wherever it was used, It can find the inline and things would fix up.

**Comment by real-or-random** [comment]
Awesome. Can you also squash the first three commits together?

**Comment by Harshil-Jani** [comment]
> Awesome. Can you also squash the first three commits together?

Done !! Thanks for reviewing.

File: examples/schnorr.c [hunk 1]
```diff
@@ -140,9 +140,8 @@ int main(void) {
      * example through "out of bounds" array access (see Heartbleed), Or the OS
      * swapping them to disk. Hence, we overwrite the secret key buffer with zeros.
      *
-     * TODO: Prevent these writes from being optimized out, as any good compiler
+     * Here we are preventing these writes from being optimized out, as any good compiler
      * will remove any writes that aren't used. */
-    memset(seckey, 0, sizeof(seckey));
-
+    secure_erase(seckey,sizeof(seckey));
```
Comment (real-or-random) [inline examples/schnorr.c:145]:
```suggestion
    secure_erase(seckey, sizeof(seckey));
```

And the same for the other calls to `secure_erase`.

**Review by real-or-random — COMMENTED** [review]
ACK with the comment addressed

That was really the last comment I have.

File: examples/schnorr.c (hunk 1 above)
Comment (Harshil-Jani) [reply examples/schnorr.c]: > That was really the last comment I have.

Oh that's totally fine. Even I am sceptical about the coding standards for established projects :) 
Done. Thanks for suggesting!!

**Review by real-or-random — APPROVED** [review]
utACK 5660c137552c657da5265691dea0fb10faae6a76

**Pull request merged**